### ISSUE TITLE
Int 8 calibration accuracy is affected by calibration batch size

### ISSUE THREAD

**Issue by EFanZh** [NVIDIA/TensorRT#774]
## Description

When using int 8 calibration, different calibration batch sizes affects the result CUDA engine accuracy.

## Environment

**TensorRT Version**: 7.1.3.4
**GPU Type**: GeForce GT 730
**Nvidia Driver Version**: 450.66
**CUDA Version**: 11.0.3
**CUDNN Version**: 8.0.2.39
**Operating System + Version**: Arch Linux
**Python Version (if applicable)**: 3.6
**TensorFlow Version (if applicable)**: 2.3.0
**PyTorch Version (if applicable)**: N/A
**Baremetal or Container (which commit + image + tag)**: Baremetal

## Relevant Files

This is the script to reproduce the problem: [int8-calibration.txt](https://github.com/NVIDIA/TensorRT/files/5198914/int8-calibration.txt).

## Steps To Reproduce

1. Install the following Python packages:
   1. `pycuda`
   2. `tensorflow`
   3. `tensorrt`
   4. `tf2onnx`
2. Run the provided script with Python.
3. Observe the accuracy difference with different calibration batch sizes.

I noticed that calibrating with batch size 1 consistently leads to lower accuracy and significantly longer calibration time than other batch sizes.

**Comment by ttyio** [comment]
Hello @EFanZh , thanks for reporting

Yes, your observation is right. The calibration pass try to extract the intermediate tensor data distribution statistics. We will build a histogram in the calibration stage, and this step is done on-the-fly, we have to expand the histogram when we observe a new absolute max in the incoming new calibration batch. As a result, we change the behavior of expanding the histogram when calibration data is provide in a different batch size. We need carefully choose calibration batch size as well as other hyper parameter in a trial-and-error way.

Hope this detail can help understand the non-determinism you observed when choose different batch size.

**Comment by ttyio** [comment]
Close since no response for more than 2 weeks, please reopen if you still have question, thanks!

**Comment by mfoglio** [comment]
So is it better to use a larger batch size for calibration or does it lead to different results but not necessarily better? What's the advice? Should I re-perform the calibration step for different batch sizes?

**Comment by ttyio** [comment]
Hello @mfoglio , the batch size impact on calibration is in the process of expanding of histogram (because each time the histogram is generated only on the current batch data). I can only tell the result sometimes different on different batch size, but it is might not as important as other tune parameter. See https://arxiv.org/abs/2004.09602, we have more detail here. Thanks!
